geniza/corpus/ja.py

```diff
@@ -79,18 +79,89 @@ def arabic_to_ja(text):
     return re.sub(re_he_final_letters, lambda m: he_final_letters[m.group(0)], text)
 
 
+# def locate_quoted_phrases(text):
+#     import re
+#     return [m.span() for m in re.finditer(r'"[^"]*"', text)]
+
+# def extract_quoted_phrases(text):
+#     return [text[m[0]:m[1]] for m in locate_quoted_phrases(text)]
+
+# def extract_arabic_phrases(text):
+#     return [
+#         phrase for phrase in extract_quoted_phrases(text) if contains_arabic(phrase)
+#     ]
+
+
+# def extract_arabic_words_outside_phrases(text):
+#     arabic_phrases = extract_arabic_phrases(text)
+
+#     # remove those arabic phrases from original string
+#     for phrase in arabic_phrases:
+#         text = text.replace(phrase, "")
+
+#     # get remaining arabic words
+#     arabic_words = extract_arabic_words(text)
+
+#     # return words
+#     return arabic_words
+
+
+# def extract_arabic_words(text):
+#     return re_AR_letters.findall(text)
+
+
+def tokenize_words_and_phrases(text):
+    import re
+
+    return [x for x in re.findall(r'"[^"]*"|[\w]+', text) if x]
+
+
+def term_is_phrase(text):
+    text = text.strip()
+    return text and " " in text and text[0] == '"' and text[-1] == '"'
+
+
+def arabic_or_ja_allowing_phrases(text, boost=True):
+    # replace arabic phrases
+    new_terms = []
+
+    for word_or_phrase in tokenize_words_and_phrases(text):
+        judeo_arabic_equivalent = arabic_to_ja(word_or_phrase)
+        if contains_arabic(word_or_phrase):
+            if term_is_phrase(word_or_phrase):
+                new_term = f"({judeo_arabic_equivalent} OR {word_or_phrase})"
+            else:
+                new_term = f"({judeo_arabic_equivalent}|{word_or_phrase})"
```

Comment:
Seeing this in context, it seems obvious that we should be using the same syntax for both cases — they are equivalent, no need to complicate the code. Let's stick with the shorter format: `(term1|term2)`

Reply:
You're right, that's much cleaner. Here's a streamlined version of the code:

```python
def arabic_or_ja_allowing_phrases2(text, boost=True):
    new_terms = [
        (
            f"({arabic_to_ja(word_or_phrase)}|{word_or_phrase})"
            if contains_arabic(word_or_phrase)
            else word_or_phrase
        )
        for word_or_phrase in tokenize_words_and_phrases(text)
    ]
    new_text = ' '.join(new_terms)
    new_text = new_text.replace('))',')').replace('((','(')
    return new_text
```